### ISSUE TITLE
Request for checkpoint conversion script from Megatron-LM to modeling_mole.py format

### ISSUE THREAD

**Issue by linfeng93** [JieShibo/MoLE#9]
Hi,

Following your suggestions, I tried to reproduce the MoLE-160M-4E results using the code in 

> https://github.com/JieShibo/MoLE/tree/main/pretrain

with `Megatron-LM core_r0.6.0`.

I used the same training data (the 100B-token Pile) and hyperparameters as described in the link above. During training, the loss decreased from 10+ to around 2.4, which seems reasonable.

However, after training I noticed that the parameter names in the Megatron-LM checkpoint are completely different from the parameter names expected by `modeling_mole.py`.

For example, the Megatron-LM checkpoint has keys like:

```
dict_keys([
    'decoder.final_layernorm.weight',
    'decoder.layers.0.mlp.experts.weight1',
    'decoder.layers.0.mlp.experts.weight2',
    'decoder.layers.0.mlp.linear_fc1.layer_norm_weight',
    'decoder.layers.0.mlp.linear_fc1.weight',
    'decoder.layers.0.mlp.linear_fc2.weight',
    'decoder.layers.0.mlp.moe_router.layer_norm_weight',
    'decoder.layers.0.mlp.moe_router.weight',
    'decoder.layers.0.self_attention.linear_proj.weight',
    'decoder.layers.0.self_attention.linear_qkv.layer_norm_weight',
    'decoder.layers.0.self_attention.linear_qkv.weight',
    ...
])
```

While `modeling_mole.py` expects something like:

```
dict_keys([
    'model.embed_tokens.weight',
    'model.layers.0.self_attn.q_proj.weight',
    'model.layers.0.self_attn.k_proj.weight',
    'model.layers.0.self_attn.v_proj.weight',
    ...
])
```

Because of this mismatch, I can't directly load the Megatron-LM checkpoint into the modeling_mole.py model definition.

**Could you please provide a conversion script (or some guidance) to map Megatron-LM checkpoints to the parameter naming/format expected by modeling_mole.py?**

This would be very helpful for reproducing the MoLE-160M-4E results.

Thanks!

**Comment by linfeng93** [comment]
As a follow-up, I tried to work around the checkpoint format mismatch by doing a two-step conversion from Megatron-LM's `model_optim_rng.pt` to the format expected by `modeling_mole.py`.

### 1. Step 1: Convert `model_optim_rng.pt` → `safetensors`

I first convert Megatron's `model_optim_rng.pt` into a safetensors file by extracting only the `model` state dict and keeping all tensor keys as-is:

```
ckpt = torch.load(input_path, map_location="cpu", weights_only=False)

if "model" not in ckpt:
    raise RuntimeError("This .pt file does not contain 'model' key!")
model_state = ckpt["model"]

tensor_state = {}
for k, v in model_state.items():
    if isinstance(v, torch.Tensor):
        tensor_state[k] = v

save_file(tensor_state, output_path, metadata={"format": "pt"})
```

This produces a `model.safetensors` that has keys like:

> embedding.word_embeddings.weight
> decoder.final_layernorm.weight
> decoder.layers.0.mlp.experts.weight1
> decoder.layers.0.mlp.experts.weight2
> decoder.layers.0.mlp.linear_fc1.layer_norm_weight
> decoder.layers.0.mlp.linear_fc1.weight
> decoder.layers.0.mlp.linear_fc2.weight
> decoder.layers.0.mlp.moe_router.layer_norm_weight
> decoder.layers.0.mlp.moe_router.weight
> decoder.layers.0.self_attention.linear_proj.weight
> decoder.layers.0.self_attention.linear_qkv.layer_norm_weight
> decoder.layers.0.self_attention.linear_qkv.weight
> ...

### 2. Step 2: Rename and reshape to `modeling_mole.py` format

Then I load that safetensors file and remap it into the naming/shape convention expected by `modeling_mole.py`. In particular:

- **Embeddings**

`state_dict_new["model.embed_tokens.weight"] = state_dict_old["embedding.word_embeddings.weight"]`


- **Final RMSNorm / output norm**

`state_dict_new["model.norm.weight"] = state_dict_old["decoder.final_layernorm.weight"]
`

- **LM head**

`state_dict_new["lm_head.weight"] = state_dict_old["output_layer.weight"]
`

- **Model hyperparameters I assume for MoLE-160M-4E:**

```
hidden_size = 768
num_layers = 12
num_heads = 12
num_kv_heads = 12
head_dim = hidden_size // num_heads
kv_dim = num_kv_heads * head_dim
num_experts = 4
inter_size = 3072
```

For each layer `i`, I map:

**(a) Attention input layer norm**

```
state_dict_new[f"model.layers.{i}.input_layernorm.weight"] = state_dict_old[f"decoder.layers.{i}.self_attention.linear_qkv.layer_norm_weight"]
```

**(b) QKV projection split**

I assume Megatron's `linear_qkv.weight` is `[hidden_size + 2 * kv_dim, hidden_size]` and I split it into `Q`, `K`, `V`:

```
W_qkv = state_dict_old[f"decoder.layers.{i}.self_attention.linear_qkv.weight"]
q_dim = hidden_size
k_dim = kv_dim

W_q = W_qkv[:q_dim, :]
W_k = W_qkv[q_dim:q_dim + k_dim, :]
W_v = W_qkv[q_dim + k_dim:, :]

state_dict_new[f"model.layers.{i}.self_attn.q_proj.weight"] = W_q
state_dict_new[f"model.layers.{i}.self_attn.k_proj.weight"] = W_k
state_dict_new[f"model.layers.{i}.self_attn.v_proj.weight"] = W_v
```

**(c) Attention output projection**

```
state_dict_new[f"model.layers.{i}.self_attn.o_proj.weight"] =  state_dict_old[f"decoder.layers.{i}.self_attention.linear_proj.weight"]
```

**(d) MLP post-attention layer norm**

```
state_dict_new[f"model.layers.{i}.post_attention_layernorm.weight"] = state_dict_old[f"decoder.layers.{i}.mlp.linear_fc1.layer_norm_weight"]
```

**(e) MLP dense layers**

```
state_dict_new[f"model.layers.{i}.mlp.up_proj.weight"] = state_dict_old[f"decoder.layers.{i}.mlp.linear_fc1.weight"]
state_dict_new[f"model.layers.{i}.mlp.down_proj.weight"] = state_dict_old[f"decoder.layers.{i}.mlp.linear_fc2.weight"]
```

**(f) MoE router + expert layernorm**

```
state_dict_new[f"model.layers.{i}.router.weight"] = state_dict_old[f"decoder.layers.{i}.mlp.moe_router.weight"]
state_dict_new[f"model.layers.{i}.expert_layernorm.weight"] = state_dict_old[f"decoder.layers.{i}.mlp.moe_router.layer_norm_weight"]
```

**(g) MoE experts**

Here I assume:

```
w1 = state_dict_old[f"decoder.layers.{i}.mlp.experts.weight1"]  # [H, E*I]
w2 = state_dict_old[f"decoder.layers.{i}.mlp.experts.weight2"]  # [E*I, H]

H = hidden_size  # 768
E = num_experts  # 4
I = inter_size   # 3072

w1_view = w1.reshape(E, H, I)  # [4, 768, 3072]
w2_view = w2.reshape(E, I, H)  # [4, 3072, 768]

for e in range(E):
    state_dict_new[f"model.layers.{i}.experts.{e}.up_proj.weight"] = w1_view[e].transpose(1, 0).contiguous()   # [I, H]
    state_dict_new[f"model.layers.{i}.experts.{e}.down_proj.weight"] = w2_view[e].transpose(1, 0).contiguous()   # [H, I]
```

After this, I save `state_dict_new` as a new safetensors file and load it with the `modeling_mole.py` implementation. **The model loads successfully without key/shape errors.**

### Problem: generation is gibberish

Despite the seemingly successful conversion and the fact that the Megatron model's training loss dropped from ~10 to ~2.4 on the 100B-token Pile, text generation looks wrong. For example, with a simple prompt like:

> "hello, I am"

the model **produces mostly garbage / non-English tokens**, not coherent English text. So it looks like something is still off in the conversion or configuration, even though shapes and keys match.

### Questions / requests for guidance

- **Is my interpretation of the Megatron QKV layout correct?**

- **Is the MoE experts reshaping / transposing logic correct?**
    Specifically:
    ```
    experts.weight1 is [H, E * I], reshaped to [E, H, I], transposed to [I, H] for each expert's up_proj.
    experts.weight2 is [E * I, H], reshaped to [E, I, H], transposed to [H, I] for each expert's down_proj.
    ```
    Is this consistent with how your MoLE implementation expects expert weights?

Thank you very much for your help🙇‍!

**Comment by JieShibo** [comment]
The file for the checkpoint conversion is no longer available, and it’s indeed easy to make mistakes when converting the qkv projection and expert weights. Here, it’s recommended to infer the dimension order of the two based on
[https://github.com/JieShibo/MoLE/blob/9eaa5c79081c4fcecdfa610ed9fec4ce0b1f5d5c/pretrain/megatron/core/transformer/attention.py#L497](https://github.com/JieShibo/MoLE/blob/9eaa5c79081c4fcecdfa610ed9fec4ce0b1f5d5c/pretrain/megatron/core/transformer/attention.py#L497)
and
[https://github.com/JieShibo/MoLE/blob/9eaa5c79081c4fcecdfa610ed9fec4ce0b1f5d5c/pretrain/megatron/core/transformer/moe/experts.py#L142](https://github.com/JieShibo/MoLE/blob/9eaa5c79081c4fcecdfa610ed9fec4ce0b1f5d5c/pretrain/megatron/core/transformer/moe/experts.py#L142).

**Comment by linfeng93** [comment]
Thank you for pointing out the subtlety when converting the QKV projection — I really appreciate it. I checked my script again and confirmed that I did make a mistake there. The previously posted:

```
W_q = W_qkv[:q_dim, :]
W_k = W_qkv[q_dim:q_dim + k_dim, :]
W_v = W_qkv[q_dim + k_dim:, :]
```

should be corrected to:

```
W_q = torch.cat([W_qkv[i * (head_dim * 3): i * (head_dim * 3) + head_dim] for i in range(num_heads)], dim=0)
W_k = torch.cat([W_qkv[i * (head_dim * 3) + head_dim : i * (head_dim * 3) + 2 * head_dim] for i in range(num_heads)], dim=0)
W_v = torch.cat([W_qkv[i * (head_dim * 3) + 2 * head_dim : i * (head_dim * 3) + 3 * head_dim] for i in range(num_heads)], dim=0)
```

After fixing this, the original issue is now resolved.

However, while debugging, I found a separate bug in the training code. I’ve documented the details here: #10 .

Thanks again for the helpful guidance!
